practices/appeng/adrs/0002-distributed-tracing.md

```diff
@@ -0,0 +1,140 @@
+- improve time to market
+- facilitates excellent cross-team communication and cooperation
+
+Here are several example scenarios or questions that distributed tracing can
+help answer.
+
+- As a new engineer on the team, I want to understand how many separate systems
+  are involved when a certain user type logs in and the first page is rendered.
+- As an operations engineer, I want to know how many SQL queries are executed
+  for a given endpoint or interaction.
+- As a product manager, I want to know if a new feature is being used by a
+  certain cohort of users on a regular basis.
+- As an engineer, I want to prove that an optimization I wrote is effective
+  in a production environment.
+- As a load tester, after I have shown that a problem exists, I want to
+  understand how the system is interacting so I can debug and fix the issue.
+
+### ADR Goals and Anti-goals
+
+- Goal: Choose which set of libraries to use at callsites (across programming
+  languages) within the codebase, which will be used to generate
+  distributed tracing data
+- Anti-goal: Committing to a specific "backend", i.e. platform or service for
+  gathering, exploring, and displaying trace information
+  - Sub-goal: Leaving open as many options as possible for the backend
+
+## Decision
+
+## Use OpenTelemetry
+
+- OpenTelemetry is an emerging industry standard
+  - vendors find benefit of being in the OpenTelemetry ecosystem because they
+    no longer have to create or support instrumentation libraries in an ever
+    growing array of languages, i.e. as soon as language library exists for
+    OpenTelemetry, the vendors automatically become available to support that
+    given language.
+- OpenTelemetry is vendor agnostic
+  - tracing information can be sent to hosted services (e.g. Honeycomb.io, AWS
+    X-Ray, etc) or self-hosted Open Source implementations (e.g. Zipkin, Jaeger,
+    etc)
+  - if left unconfigured, OpenTelemetry instrumentation calls default to
+    lightweight/noop executions
+- OpenTelemetry has well-maintained libraries for the languages used in the
+  layers of most Truss projects
+  - i.e. Go (back-end); JavaScript (front-end); Python (load testing); etc
+- Easily swappable back-ends
+  - e.g. could choose a local Docker version of OpenZipkin for an all-local
+    development environment
+  - e.g. can use Honeycomb.io in the experimental commercial-cloud hosted
+    environment
+  - e.g. can swap in AWS X-Ray for use in GovCloud hosted environments
+- Cons
+  - as an abstraction layer, OpenTelemetry may prohibit usage of vendor-
+    specific capabilities
+  - some OpenTelemetry libraries and tools may trail their vendor-supported
+    counterparts
+  - instrumentation for tracing may be a vector for performance overhead
+
+## Why is this Applicable to the Practice as a Whole
+
+TODO: What makes this decision generally applicable
+to Truss projects?
+
+## When to Not Implement This Decision
```

Comment:
* Don't use OpenTelemetry if the ecosystem you are working in doesn't have a mature library

Reply:
I like this criteria. My understanding is that it's possible that this could change over time if a contract scales and the project gets bigger. 

Is it a good idea to include a better breakdown of what the thresholds look like for implementing tracing?

It can be difficult to add tracing after the fact for a mature project. How can this ADR address having an MVP that is friendly to building out OpenTelemetry if the project scales up and eventually needs it?

I speak from experience on MilMove that implementing OpenTelemetry came at the cost of a large number of PRs that were difficult to review.

I'm musing on wording here that explains how we can possibly architect a solution that is OpenTelemetry-friendly even though it hits the points you made of not needing to implement it.

Reply:
Just a couple points here:

- This ADR doesn't supersede the need for implementation documentation, which should have more of this information
- Hopefully the standardization of this makes it easier for all projects (including small ones) to quickly implement, since we have the in house knowledge